### DISCUSSION TITLE
Custom mouse behavior

### DISCUSSION THREAD

**Discussion by lchauvin** [Kitware/VolView#573] [Q&A]
Hello,

I would like to create some custom behavior for mouse events.
I want for example to have right click to activate / deactivate the crosshair (right click to activate, stay press, move, then release to deactivate), left click for painting, and middle button for pan / zoom.

What would be the best way to do that? Use the current tools and tweak the mouse behaviors, or create a new tool, that will encompass all of these features, and set its own behavior?

Thank you.

**Comment by lchauvin** [comment]
What do you think would be the best approach ?

**Reply by PaulHax** [reply]
I'd start by registering `capture` option handler for the mouse events that gets called before the tools get the event.  This handler will chang the selected tools.  Some relevent code in this PR https://github.com/Kitware/VolView/pull/441/files and the old issue https://github.com/Kitware/VolView/issues/422

**Reply by lchauvin** [reply]
where do you suggest I capture this ? In `action.ts` ?

**Reply by PaulHax** [reply]
In the view 2d component so that this tool swiching handler gets called before the tool handlers.

**Reply by lchauvin** [reply]
I added this code in `VtkTwoView.vue`:

```
    // Handle mouse events before widgets
    const interactor = computed(() => viewProxy.value?.getInteractor());
    onVTKEvent(
      interactor,
      'onLeftButtonPress',
      (event: any) => {
        toolStore.setCurrentTool(Tools.Crosshairs);
      },

    );
    onVTKEvent(
      interactor,
      'onLeftButtonRelease',
      (event: any) => {
        toolStore.setCurrentTool(Tools.Paint);
      },
```

And it kind of works. However, when I set the crosshair tool as active, it shows the crosshair, but at the last location it was set, and only update it when I start moving the cursor.
It's a bit of a detail but everytime I reactivate the crosshair, it 'jumps' from the last location it was active to the new position of the cursor.

Is there a way to set it to the current location of the cursor when activating it? Shouldn't it be the default behavior?

**Reply by lchauvin** [reply]
The same thing kind of happen for the paint widget. When I activate it, the brush is not visible, it only becomes visible when I start moving the cursor.

**Reply by PaulHax** [reply]
Great bug finds!  Need to fix these tools to "update" with mouse XY when they are activated, (if mouse is within view?) 

Now the fun begins!

**Reply by lchauvin** [reply]
I created an issue: https://github.com/Kitware/VolView/issues/574

I will try to fix it and create a MR.

**Reply by PaulHax** [reply]
Not sure offhand.  I would try to propigate the initial event down to the tool.  Then the tool runs its logic as usual as if it was ready and wating for the event, as if it was already "active".

**Reply by lchauvin** [reply]
I believe in `crosshairs.ts`:

```
  // update the position
  handle.onModified(() => {
    const origin = handle.getOrigin();
    if (origin) {
      position.value = origin;
    }
  });

  function activateTool() {
    widgetState.setPlaced(false);
    active.value = true;
    return true;
  }
```

`activateTool` should trigger the `handle.onModified` to update the crosshair. However `handle.getOrigin()` is `[0,0,0]` by default. So `activateTool` will need to get the current position of the cursor in the view (if the cursor is in a view).

Does that makes sense? Is it the right way of doing this?

**Reply by PaulHax** [reply]
That makes sense, but I'm hoping there is an easier approach by just letting the first mouse event pass through and trigger the lower level "vtk" widget for the tool (ie `src/vtk/CrosshairsWidget`).  So hoping that only thing to do is make the low level vkt widget sensitive to the event and call the code it usualy calls if the tool was already selected.

**Reply by lchauvin** [reply]
The problem is that is not only linked to the click event.

If you open volview, load an image, move your cursor over a 2D view, and press 'c' to switch to crosshair, and not move the cursor, the crosshair doesn't appear. It seems to be more like an activation issue of the tool. When activating, it should update the rendering (or call the `modified` event, that would update the rendering).

**Reply by lchauvin** [reply]
I'm still stuck with this issue.

I'm not sure which part should be in charge of updating the crosshair when activated. Should it be the `VtkTwoView` that detect when the crosshair is activated, and if in it's view, update the position? Should it be the crosshair that should find it's position in `VtkTwoView` (if in it) when activated?

**Reply by PaulHax** [reply]
That sounds like a good approach.  Some code assosated with `VtkTwoView` needs to cache the mouse position (as keyboard event does not have it.) Then on `c`, trigger some update on the crosshairs widget.  

Could be nice to get this fucntion to fire which converts from mouse coordinates to "world space":
https://github.com/Kitware/VolView/blob/main/src/vtk/CrosshairsWidget/behavior.ts#L48
Perhaps by asking the VTK.js WidgetManager class to process a synthesized mouse event from the cached mouse position.  

(Also, looks like there is a bug if I press `c` twice...)

**Reply by lchauvin** [reply]
Yes, I've been trying to trigger this function with a mouse event, but couldn't do it. I couldn't find a way to call it directly from `VtkTwoView` either. I'm not sure how to access the behavior from the WidgetManager (sorry if it's trivial, I'm not so familiar with the code yet).

**Reply by PaulHax** [reply]
I'm learning too =)  

This function should a trigger the crosshairs behaivor widget: 
https://github.com/Kitware/vtk-js/blob/master/Sources/Widgets/Core/WidgetManager/index.js#L186

Looks like the vtk interactor calls it so maybe we need to poke an interactor somehow:
https://github.com/Kitware/vtk-js/blob/master/Sources/Widgets/Core/WidgetManager/index.js#L286-L289

Looks like `View` structures have an handle to interactors:
https://github.com/Kitware/VolView/blob/main/src/components/tools/paint/PaintWidget2D.vue#L134

I'm not familar with vtk.js interactors.  =(

**Reply by floryst** [reply]
Just like Paul mentioned, there needs to be a "caching" behavior for the mouse position that then gets passed to the initialization of the tool. This should most likely be done in the vtkCrosshairsWidget behavior code.

**Reply by lchauvin** [reply]
I also noticed this comment in vtkCrosshairWidget behavior:

```
  publicAPI.handleMouseMove = (callData: any) => {
    // technically need to call requestAnimation during
    // the initial phase of placing the crosshairs,
    // but that's not needed since no VTK object
    // is actually being rendered.
```

I'm not sure if it's related to the same issue or not.

**Comment by floryst** [comment]
Having custom mouse behavior is a generally good thing to have! While I don't have an exact pathway for this, I have two thoughts on it that I'll revisit later.
- The vtk.js widgets should support configurable actions directly. A concrete example of this would be binding crosshair interactions to left mouse or right mouse buttons.
- A vtkInteractorObserver middleware that supports configurable button translations. A concrete example of this would be to translate right buttons to left button events, passing on the left button event to downstream observers.
